core/src/test/java/io/atomix/core/AtomixTest.java

```diff
@@ -246,6 +250,152 @@ public void testRoleChangedListener() throws Exception {
+        || nodeRoles.get(2).size() < 3) {
+      Thread.sleep(100L);
+    }
+
+    final Map<Integer, RaftServer.Role> expectedNodeOneRoles = new HashMap<>();
+    expectedNodeOneRoles.put(1, Role.LEADER);
+    expectedNodeOneRoles.put(2, Role.FOLLOWER);
+    expectedNodeOneRoles.put(3, Role.FOLLOWER);
+    assertEquals(expectedNodeOneRoles, nodeRoles.get(0));
+
+    final Map<Integer, RaftServer.Role> expectedNodeTwoRoles = new HashMap<>();
+    expectedNodeTwoRoles.put(1, Role.FOLLOWER);
+    expectedNodeTwoRoles.put(2, Role.LEADER);
+    expectedNodeTwoRoles.put(3, Role.FOLLOWER);
+    assertEquals(expectedNodeTwoRoles, nodeRoles.get(1));
+
+    final Map<Integer, RaftServer.Role> expectedNodeThreeRoles = new HashMap<>();
+    expectedNodeThreeRoles.put(1, Role.FOLLOWER);
+    expectedNodeThreeRoles.put(2, Role.FOLLOWER);
+    expectedNodeThreeRoles.put(3, Role.LEADER);
+    assertEquals(expectedNodeThreeRoles, nodeRoles.get(2));
+  }
+
+  @Test
+  public void testAtomixBootstrapPartitionsAndRestartingNode() throws Exception {
+    // given
+    // Partitions \ Nodes
+    //      \   0  1  2
+    //    0     L  F  F
+    //    1     F  L  F
+    //    2     F  F  L
+
+    final List<Map<Integer, RaftServer.Role>> nodeRoles = new ArrayList<>();
+    nodeRoles.add(new HashMap<>());
+    nodeRoles.add(new HashMap<>());
+    nodeRoles.add(new HashMap<>());
+
+    final List<Integer> members = Arrays.asList(1, 2, 3);
+
+    final int firstNodeId = 1;
+    final CompletableFuture<Atomix> nodeOneFuture = setupAtomixWithListener(firstNodeId, members, nodeRoles);
+
+    final int secondNodeId = 2;
+    final CompletableFuture<Atomix> nodeTwoFuture = setupAtomixWithListener(secondNodeId, members, nodeRoles);
+
+    final int thirdNodeId = 3;
+    final CompletableFuture<Atomix> nodeThreeFuture = setupAtomixWithListener(thirdNodeId, members, nodeRoles);
+
+    // then
+    CompletableFuture.allOf(nodeOneFuture, nodeTwoFuture, nodeThreeFuture).join();
+
+    while (nodeRoles.get(0).size() < 3
+        || nodeRoles.get(1).size() < 3
+        || nodeRoles.get(2).size() < 3) {
+      Thread.sleep(100L);
+    }
+
+    // when
+    final Atomix atomix = nodeTwoFuture.get();
+    atomix.stop().join();
+    nodeRoles.get(1).clear();
+    final CompletableFuture<Atomix> nodeTwoSecondFuture = setupAtomixWithListener(secondNodeId,
+        members, nodeRoles);
+
+    nodeTwoSecondFuture.join();
+    while (nodeRoles.get(0).size() < 3
+        || nodeRoles.get(1).size() < 3
+        || nodeRoles.get(2).size() < 3) {
+      Thread.sleep(100L);
+    }
+
+    // then
+    final long nodeOneLeaderCount = nodeRoles.get(0).values().stream().filter(r -> r == Role.LEADER).count();
+    final long nodeTwoLeaderCount = nodeRoles.get(1).values().stream().filter(r -> r == Role.LEADER).count();
+    final long nodeThreeLeaderCount = nodeRoles.get(2).values().stream().filter(r -> r == Role.LEADER).count();
+
+    assertTrue(nodeOneLeaderCount == 2 || nodeThreeLeaderCount == 2);
```

Comment:
Also flaky (maybe?), sometimes this fails with false - maybe if an election occurs when we bring nodes down we won't have the expected count?